### PR TITLE
Feat/core entities

### PR DESCRIPTION
**Pull request by stawuah**
A  fresh pair of eyes is always valuable to catch anything I might have missed. hehe

### CODE REVIEW

File: controllers/student.controller.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        let newUser;
+        let regNumber;
+
+        try {
+
+            this.registrationUtils.validateStudentInfo(userData, userType,);
+            const { email, role, } = await this.registrationUtils.prepareData(userData);
+            switch (userType) {
+                case 'student':
+                    regNumber = generateRegNumber()
```
Comment (RWUBAKWANAYO) [inline controllers/student.controller.js]: Please validate regNumber before we register it to the collection

**Review by RWUBAKWANAYO — CHANGES_REQUESTED** [review]

File: controllers/staff.controllers.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    findAllStaff = async (req, res, next) => {
+        try {
+            const staff = await this.staff.getAllStaff();
+
+            if (staff.length === 0) {
+                return res.status(404).json({ message: 'No students found .' })
+            }
+            res.status(200).json({ Allstaff: staff })
+        } catch (error) {
+            console.error(`Error retrieving all students `, error);
```
Comment (RWUBAKWANAYO) [inline controllers/staff.controllers.js]: Please change this error message

File: routes/staff.routes.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+const { validateToken } = require('../middlewares/validations');
+const StaffController = require('../controllers/staff.controllers');
+const { authorize } = require('../middlewares/authorize');
+const staffController = new StaffController();
+const router = express.Router();
+
+
+
+router.get('/staff/students', validateToken, authorize(['readStudents']), staffController.findAllStaff);
+router.delete('/no/staff', validateToken, authorize(['deleteStaff']), staffController.deleteStaffById);
```
Comment (RWUBAKWANAYO) [inline routes/staff.routes.js]: How are we going to interact with these routes without passing Ids?

File: routes/staff.routes.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+const express = require('express');
+const { validateToken } = require('../middlewares/validations');
+const StaffController = require('../controllers/staff.controllers');
+const { authorize } = require('../middlewares/authorize');
+const staffController = new StaffController();
+const router = express.Router();
+
+
+
+router.get('/staff/students', validateToken, authorize(['readStudents']), staffController.findAllStaff);
```
Comment (RWUBAKWANAYO) [inline routes/staff.routes.js]: We arleady created this route inside `students.authroutes.js`

**Review by RWUBAKWANAYO — CHANGES_REQUESTED** [review]

File: routes/staff.routes.js (hunk 4 above)
Comment (stawuah) [reply routes/staff.routes.js]: fixed,

File: routes/staff.routes.js (hunk 3 above)
Comment (stawuah) [reply routes/staff.routes.js]: Fixed 
```suggestion
router.delete('/no/staff/:staff_id', validateToken, authorize(['deleteStaff']), staffController.deleteStaffById);
```

File: routes/staff.routes.js (hunk 4 above)
Comment (stawuah) [reply routes/staff.routes.js]:
```suggestion
router.get('/all/staff', validateToken, authorize(['readStaff']), staffController.findAllStaff);
```

File: controllers/student.controller.js (hunk 1 above)
Comment (stawuah) [reply controllers/student.controller.js]: Fixed 
```suggestion
                    regNumber = generateRegNumber() 
```

Comment (stawuah) [reply controllers/student.controller.js]: Registration Validation fixed , no duplicate of reg_numbers and an extra validation to verify 11 digit reg_number before registration and these implementation were done in the model and the ./utils/utils.registrationNumber.js

File: controllers/staff.controllers.js (hunk 2 above)
Comment (stawuah) [reply controllers/staff.controllers.js]: Error message fixed

**Review by RWUBAKWANAYO — COMMENTED** [review]
Please change the name of `routes/calss.route.js` file to `routes/class.route.js`

**Review by RWUBAKWANAYO — APPROVED** [review]

**Review by stawuah — COMMENTED** [review]
changes done

**Review by stawuah — COMMENTED** [review]
fixed

**Review by stawuah — COMMENTED** [review]
Looks good to me  , @RWUBAKWANAYO

**Pull request merged**